homeassistant/components/zha/entity.py

```diff
@@ -66,20 +66,34 @@ def __init__(self, entity_data: EntityData, *args, **kwargs) -> None:
 
     @cached_property
     def name(self) -> str | UndefinedType | None:
-        """Return the name of the entity."""
+        """Return the name of the entity.
+
+        Built-in quirks have translations in HA, so those are used.
+        Custom quirks with new translation keys won't have translations.
+        For them, the fallback name should be used instead.
+        If a device class is set but no translation key,
+        the device class name is used.
+        """
         meta = self.entity_data.entity.info_object
         if meta.primary:
             self._attr_name = None
             return super().name
 
-        original_name = super().name
-
-        if original_name not in (UNDEFINED, None) or meta.fallback_name is None:
-            return original_name
-
-        # This is to allow local development and to register niche devices, since
-        # their translation_key will probably never be added to `zha/strings.json`.
-        self._attr_name = meta.fallback_name
+        # if we have a translation key or no (device class) name,
+        # use translation or fallback name
+        if meta.translation_key is not None or super().name in (UNDEFINED, None):
+            # if we have a translation for this key, use it
+            if (
+                (translation_key := self._name_translation_key) is not None
+                and translation_key in self.platform_data.platform_translations
+            ):
+                return super().name
+            # a translation won't be available for custom quirks, so use fallback_name
+            if meta.fallback_name is not None:
+                self._attr_name = meta.fallback_name
+                return super().name
+
+        # we should only reach this for the device class name at this point
         return super().name
```

Comment:
I think your version is unnecessarily complicated with the nested if statements.
This one works just as well (slightly adjusted from my original - and maybe missing comments in the right place)

```suggestion
        if (
            meta.fallback_name is not None
            and meta.translation_key is not None
            and (
                (translation_key := self._name_translation_key) is None
                or translation_key not in self.platform_data.platform_translations
            )
        ):
            self._attr_name = meta.fallback_name
        return super().name
```

Reply:
But that breaks with the one number test you had [here](https://github.com/home-assistant/core/pull/152155#issuecomment-3290790360), where the entity neither has a `device_class` nor a `translation_key`, just a `fallback_name` based on the ZCL description attribute.

I'll see if I can expand the new test to explicitly test for that as well. To fix this, the `super().name in (UNDEFINED, None)` check would need to be added properly to your condition as well.

But at that point, the "nested if version" might be more easy to understand, hence I went with that + comments. This might be something we can solve with https://github.com/zigpy/zha/issues/527 eventually.

Reply:
```suggestion
        # If we do not have a fallback_name, use default behavior
        if meta.fallback_name is None:
            return super().name

        # If we do not have a translation key, only use fallback_name
        # if device class is also missing
        if meta.translation_key is None:
            if super().name in (UNDEFINED, None):
                self._attr_name = meta.fallback_name
            return super().name

        # If we do have a translation key, only use fallback_name
        # if translation is missing (custom quirks)
        if not (
            (translation_key := self._name_translation_key) is not None
            and translation_key in self.platform_data.platform_translations
        ):
            self._attr_name = meta.fallback_name
        return super().name
```

Reply:
That looks nice and readable to me! 😄